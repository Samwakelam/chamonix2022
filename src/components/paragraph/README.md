# Paragraph

A basic HTML Paragraph element, can be set different font sizes through the `size` prop. Content is passed in via `children` prop.

## Import

```jsx
import { Paragraph } from '@gymshark/design-system';
```

# Reference

## Props

| Prop       | Required | Type        | Description                                                 |
| ---------- | -------- | ----------- | ----------------------------------------------------------- |
| `size`     | `true`   | `ENUM`      | Import `PargraphSizes` for ENUM provider, changes font size |
| `children` | `true`   | `ReactNode` | Valid react children element                                |
