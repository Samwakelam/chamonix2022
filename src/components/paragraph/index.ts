export { Paragraph } from './paragraph.component';
export { ParagraphSizes } from './paragraph.definition';
