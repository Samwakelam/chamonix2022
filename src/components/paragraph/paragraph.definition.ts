import React from 'react';

export enum ParagraphSizes {
    SMALL = 'small',
    STANDARD = 'standard',
    LARGE = 'large',
}

export type SingleConfig = {
    fontSize: string;
    lineHeight: string;
};

export interface ParagraphProps<T extends React.ElementType> {
    as?: T;
    children: React.ReactNode;
    size?: ParagraphSizes;
}

export type ParagraphStyledProps = {
    configuration: SingleConfig;
};

export type ParagraphSizesShape = {
    [key: string]: SingleConfig;
};
