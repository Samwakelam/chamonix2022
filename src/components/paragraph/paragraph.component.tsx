import React, { ComponentPropsWithoutRef } from 'react';

import { ParagraphProps, ParagraphSizes } from './paragraph.definition';
import { paragraphSizes } from './paragraph.shape';

import * as S from './paragraph.styles';

export const Paragraph = <T extends React.ElementType = 'p'>({
    children,
    size = ParagraphSizes.STANDARD,
}: ParagraphProps<T> & Omit<ComponentPropsWithoutRef<T>, keyof ParagraphProps<T>>): React.ReactElement<
    ParagraphProps<T>
> => <S.Paragraph configuration={paragraphSizes[size]}>{children}</S.Paragraph>;
