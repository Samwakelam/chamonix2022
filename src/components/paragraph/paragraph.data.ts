import { ParagraphProps, ParagraphSizes } from './paragraph.definition';

export const ParagraphData: ParagraphProps<'p'> = {
    children:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Proin eu magna lorem. Morbi scelerisque sapien sit amet mi bibendum elementum. Praesent id nibh vehicula, dapibus mauris eu, maximus ipsum. Vivamus a mauris sed justo aliquam euismod aliquet quis tortor. Quisque quis luctus nisi. Fusce nisi nunc, aliquet quis ultrices eget, lobortis nec justo. Phasellus at faucibus nisl, ut consequat leo. Curabitur hendrerit quam diam, vitae ornare tellus consequat in.',
    size: ParagraphSizes.STANDARD,
};
