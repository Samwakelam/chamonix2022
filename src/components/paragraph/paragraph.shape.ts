import { ParagraphSizesShape } from './paragraph.definition';

export const paragraphSizes: ParagraphSizesShape = {
    small: {
        fontSize: '1.4rem',
        lineHeight: '2.1rem',
    },
    standard: {
        fontSize: '1.6rem',
        lineHeight: '2.4rem',
    },
    large: {
        fontSize: '2.1rem',
        lineHeight: '3.15rem',
    },
};
