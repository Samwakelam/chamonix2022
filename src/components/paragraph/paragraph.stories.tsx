import React, { ComponentPropsWithoutRef } from 'react';
import { Meta, Story } from '@storybook/react';
import '@storybook/addon-console';

import { Paragraph as ParagraphComponent } from './paragraph.component';
import { ParagraphProps, ParagraphSizes } from './paragraph.definition';
import { ParagraphData } from './paragraph.data';

export default {
    component: ParagraphComponent,
    title: 'Atoms/Typography/Paragraph',
} as Meta;

const Paragraph: Story<
    ParagraphProps<React.ElementType> &
        Omit<ComponentPropsWithoutRef<React.ElementType>, keyof ParagraphProps<React.ElementType>>
> = ({ children, ...args }: ParagraphProps<React.ElementType>) => (
    <ParagraphComponent {...args}>{children}</ParagraphComponent>
);

export const Standard = Paragraph.bind({});

Standard.args = {
    children: ParagraphData.children,
    size: ParagraphSizes.STANDARD,
};

export const Small = Paragraph.bind({});

Small.args = {
    children: ParagraphData.children,
    size: ParagraphSizes.SMALL,
};

export const Large = Paragraph.bind({});

Large.args = {
    children: ParagraphData.children,
    size: ParagraphSizes.LARGE,
};
