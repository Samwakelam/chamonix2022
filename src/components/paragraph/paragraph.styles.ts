import styled from 'styled-components';

import { theme } from '../../theme/theme.object';

import { ParagraphStyledProps } from './paragraph.definition';

export const Paragraph = styled.p<ParagraphStyledProps>`
    margin: 0;
    font-family: ${theme.f.lato};
    font-size: ${({ configuration: { fontSize } }) => fontSize};
    line-height: ${({ configuration: { lineHeight } }) => lineHeight};
`;
