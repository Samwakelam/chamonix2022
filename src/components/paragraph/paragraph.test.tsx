import React from 'react';
import { render } from '@testing-library/react';

import { Paragraph } from './paragraph.component';
import { ParagraphData } from './paragraph.data';

describe('Paragraph Component', () => {
    const { container } = render(<Paragraph {...ParagraphData} />);

    it('should match snapshot correctly', () => {
        expect(container.firstChild).toMatchSnapshot();
    });
});
