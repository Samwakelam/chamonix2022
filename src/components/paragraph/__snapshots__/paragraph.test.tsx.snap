// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paragraph Component should match snapshot correctly 1`] = `
<p
  class="sc-bdfBQB dHdZBQ"
>
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Proin eu magna lorem. Morbi scelerisque sapien sit amet mi bibendum elementum. Praesent id nibh vehicula, dapibus mauris eu, maximus ipsum. Vivamus a mauris sed justo aliquam euismod aliquet quis tortor. Quisque quis luctus nisi. Fusce nisi nunc, aliquet quis ultrices eget, lobortis nec justo. Phasellus at faucibus nisl, ut consequat leo. Curabitur hendrerit quam diam, vitae ornare tellus consequat in.
</p>
`;
